model/generate.go

```diff
@@ -275,25 +270,160 @@ func (g *GeneratedProject) saveNewBuildsAndTasks(ctx context.Context, v *Version
 	}
 
 	activatedTasksInExistingBuilds, err := addNewTasks(ctx, activationInfo, v, p, projectRef, newTVPairsForExistingVariants,
-		existingBuilds, syncAtEndOpts, g.TaskID)
+		existingBuilds, syncAtEndOpts, g.Task.Id)
 	if err != nil {
 		return errors.Wrap(err, "adding new tasks")
 	}
 
 	activatedTasksInNewBuilds, err := addNewBuilds(ctx, activationInfo, v, p, newTVPairsForNewVariants,
-		existingBuilds, syncAtEndOpts, projectRef, g.TaskID)
+		existingBuilds, syncAtEndOpts, projectRef, g.Task.Id)
 	if err != nil {
 		return errors.Wrap(err, "adding new builds")
 	}
 
 	// only want to add dependencies to activated tasks
-	if err = addDependencies(t, append(activatedTasksInExistingBuilds, activatedTasksInNewBuilds...)); err != nil {
+	if err = g.addDependencies(append(activatedTasksInExistingBuilds, activatedTasksInNewBuilds...)); err != nil {
 		return errors.Wrap(err, "adding dependencies")
 	}
 
 	return nil
 }
 
+func (g *GeneratedProject) SimulateNewDependencyGraph(v *Version, p *Project, projectRef *ProjectRef) error {
+	existingTasksGraph, err := task.VersionDependencyGraph(g.Task.Version, false)
+	if err != nil {
+		return errors.Wrapf(err, "creating dependency graph for version '%s'", g.Task.Version)
+	}
+
+	simulatedGraph, err := g.addNewTasksToGraph(existingTasksGraph, v, p, projectRef)
+	if err != nil {
+		return errors.Wrap(err, "simulating new tasks")
+	}
+
+	if cycles := simulatedGraph.Cycles(); len(cycles) > 0 {
+		return errors.Wrapf(DependencyCycleError, "'%s'", cycles)
+	}
+
+	return nil
+}
+
+func (g *GeneratedProject) addNewTasksToGraph(graph task.DependencyGraph, v *Version, p *Project, projectRef *ProjectRef) (task.DependencyGraph, error) {
+	newTasks := g.getNewTasksWithDependencies(v, p)
+
+	taskIDs, err := getTaskIdTables(v, p, newTasks, projectRef.Identifier)
+	if err != nil {
+		return graph, errors.Wrap(err, "getting task ids")
+	}
+
+	graph = addTasksToGraph(newTasks.ExecTasks, graph, p, taskIDs)
+	return g.addDependencyEdgesToGraph(newTasks.ExecTasks, v, p, graph, taskIDs)
+}
+
+func (g *GeneratedProject) getNewTasksWithDependencies(v *Version, p *Project) TaskVariantPairs {
```

Comment:
By newTasks here do you mean just tasks that the generator is going to add? It doesn't look like that's what appendTasks does, since it just loops through everything that the project has, in which case I wonder if we actually need to query all the version tasks.

Reply:
That _is_ the intention of `newTasks`.
I'm not sure what you mean, though: `appendTasks` is iterating through the tasks in the `GeneratedProject`s build variants. Aren't these just the new tasks?

Reply:
Ah you're right; I was thinking it was looping through p which is the entire project but it uses bv which is the generated build variant.